src/core/image.js

```diff
@@ -506,6 +506,122 @@ var PDFImage = (function PDFImageClosure() {
       }
     },
 
+    /**
+     * Check whether resize image data or not
+     * @param {Number} comps Number of color components, 1 or 3 is supported.
+     * @param {Number} bpc Number of bits per component.
+     * @return {Boolean}
+     */
+    shallResizeImage: function PDFImage_shallResizeImage(comps, bpc) {
+        return (comps === 1 && (bpc === 1 || bpc === 8));
+    },
+
+    /**
+     * Resize black white image data
+     * @param {Uint8Array} imgData image data.
+     * @param {Number} scaleBits
+     * @author Ramsoft
+     */
+    resizeBWPixels: function PDFImage_resizeBWPixels(imgData, scaleBits) {
+      var newWidth = imgData.width >> scaleBits;
+      var newHeight = imgData.height >> scaleBits;
+
+      var newRowBytes = (newWidth + 7) >> 3;
+      var originalRowBytes = (imgData.width + 7) >> 3;
+      var step = 1 << scaleBits;
+      var numBytes = newHeight * newRowBytes;
+      var pixelArrayOutput = new Uint8Array(numBytes);
+      var originalRowStart;
+      var newRowStart;
+
+      for (var i = 0, y2 = 0; i < newHeight; i++, y2 += step,
+          originalRowStart = y2 * originalRowBytes,
+          newRowStart = i * newRowBytes) {
+        for (var j = 0, x2 = 0; j < newWidth; j++, x2 += step) {
+            // we want original value of pixel [x2, y2] value in original image
+            // to populate pixel [i, j] in new image
+
+            var originalColByteStart = x2 >> 3;
+            var originalColBitMask = 1 << (7 - (x2 & 7));
+            // most signifcant bit is first pixel due to Little Endian
+
+            var originalColByte =
+                imgData.data[originalRowStart + originalColByteStart];
+            var pixelValue = originalColByte & originalColBitMask;
+
+            if (pixelValue > 0) {
+              // most signifcant bit is first pixel due to Little Endian
+              var newColByteStart = j >> 3;
+              var newColBitMask = 1 << (7 - (j & 7));
+
+              var newColByte = pixelArrayOutput[newRowStart + newColByteStart];
+
+              newColByte = newColByte | newColBitMask;
+              // set pixel bit to 1
+              pixelArrayOutput[newRowStart + newColByteStart] = newColByte;
+            }
+        }
+      }
+
+      imgData.data = pixelArrayOutput;
+      imgData.width = newWidth;
+      imgData.height = newHeight;
+    },
+
+    /**
+     * Resize large resolution PDFS as to improve rendering time
+     * @param  {Uint8Array} imgData image data.
+     * @param  {Number} comps Number of color components, 1 or 3 is supported.
+     * @param  {Number} bpc Number of bits per component.
+     */
+    resizeGrayPixels: function PDFImage_resizeGrayPixels(imgData, comps, bpc) {
+      var scaleBits;
+      var h1 = imgData.height;
+      var w1 = imgData.width;
+
+      // Reduce the pixel dimensions to around 2000.
+      if ((h1 > 15000) || (w1 > 15000)) {
```

Comment:
https://github.com/mozilla/pdf.js/pull/6717#discussion_r46593022 still seems to be unresolved, since there are no comment explaining *how* these numbers were chosen.